pkg/apiserver/debugapi/endpoint/endpoint.go

```diff
@@ -29,15 +30,23 @@ var (
 	ErrInvalidParam         = ErrNS.NewType("invalid_parameter")
 )
 
+type APIModelPreHooks func(req *Request, data map[string]string, m *APIModel) error
+type APIModelPostHooks func(req *Request, path Values, query Values, m *APIModel) error
+
 type APIModel struct {
 	ID          string         `json:"id"`
 	Component   model.NodeKind `json:"component"`
 	Path        string         `json:"path"`
 	Method      Method         `json:"method"`
 	PathParams  []APIParam     `json:"path_params"`  // e.g. /stats/dump/{db}/{table} -> db, table
 	QueryParams []APIParam     `json:"query_params"` // e.g. /debug/pprof?seconds=1 -> seconds
+
+	PreHooks  []APIModelPreHooks  `json:"-"`
+	PostHooks []APIModelPostHooks `json:"-"`
 }
 
+// Transformers execution order:
+// APIParam.PreModelTransformer -> APIParamModel.PreTransformer -> required check -> APIParamModel.Transformer -> APIParam.PostModelTransfomer
```

Comment:
Why do we have so many transformers :/ Can we make them simple?

Reply:
Yeah, it's too complicated. Maybe I can design it as middleware. Let me try.